### ISSUE TITLE
Idea: introduce abstraction for computeds with arguments

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx-utils#184]
See https://github.com/mobxjs/mobx/issues/1388#issuecomment-472748989 and https://github.com/mobxjs/mobx/issues/1388#issuecomment-472751685. This could potentially replace `createTransformer` as well.

Signature:
`memoizedComputed<T>(expression: T, invalidationStrategy?: "auto" | "never" | number = "auto"): T`

strategies:
* `"auto"`: remove entries from the memoization table as soon as they become `unobserved`
* `"never"`: we don't care, cache forever (that is, until the computeds and all things they depend on are GC-ed)
* `number`: remove entries after they have not been hit a the specified amount of time. Probably not needed in a first versoin.

Example:

```typescript
class Todos {
  @observable todos = []
  
  private todosByUserCache = new Map()

  getAllTodosByUser = memoizedComputed((userId) => {
    if (this.todosByUserCache.has(userId))
      return this.todosByUserCache.get(userId).get()
      
    const computedFilter = computed(() => this.todos.filter(todo => todo.user === userId))
    this.todosByUserCache.set(userId, computedFilter)
    return todosByUserCache.get()
  })
}

const todos = new Todos()
const myTodos: Todo[] = todos.getAllTodosByUser("17")
```

**Comment by bourquep** [comment]
Woah, it's funny how you opened this issue 5 hours ago and my coworker and I were trying to figure out how we'd solve this very issue 10 hours ago 🤓

**Comment by urugator** [comment]
Does `"never"` pass `keepAlive: true` to the underlying `computed` ? I think it would make sense? Then maybe `computed` API could be preserved (without `number` support)... and perhaps the user should be able to provide other options like `equals`...?

**Comment by mweststrate** [comment]
Good point, yes I think so. Didn't think it that far through :)

**Comment by dreamair** [comment]
For one-object-argument functions (a User object instead of just the userId in your example) a WeakMap could help automatically cleaning up entries with the argument-object being GCed. It would be like an "extension" to the User indirectly holding a computed member for all his todos without depending on them.
I don't see this working for multiple arguments or primitives though.

**Comment by mweststrate** [comment]
Released as 5.4.0!
